File: benchmarks/canvas/draw_shapes.f
// claude.md #103: the canvas benchmark, Festina half.
//
// A frame of mixed 2D drawing -- filled rectangles, circles and text,
// with the fill colour changed between shapes so the runtime cannot
// batch them into one paint. This is what a 2D game's frame actually
// looks like, rather than one huge shape or one repeated call.
//
// Everything here paints an OFFSCREEN surface (claude.md #95); no
// display is involved and no window opens. That is deliberate and it is
// what makes the comparison against a browser canvas fair in the one
// direction that matters: both sides are measured drawing into a
// buffer, not presenting it to a screen.

int shapes = 20000

void func frame(n:int) {
    for int i = 0, i < n, i++ {
        int x = (i * 37) % 780
        int y = (i * 53) % 580
        fillStyle(i % 255, (i * 7) % 255, (i * 13) % 255)
        drawRect(x, y, 12, 9)
        drawCircle(x + 6, y + 4, 4)
    }
}

// Timed with now() around the draw loop alone, so the number is the
// same thing the browser side measures with performance.now(): the
// drawing, not the process around it. Subtracting a blank-frame run
// instead would fold in the PNG encode, and a busy image encodes far
// slower than an empty one -- which inflated Festina's figure by
// roughly a third the first time this was measured.
clearCanvas()
int start = now()
frame(shapes)
int elapsed = now() - start
log(elapsed)
log(saveCanvas('festina_canvas.png'))

File: benchmarks/canvas/draw_shapes.js
// claude.md #103: the canvas benchmark, browser half.
//
// The same frame as draw_shapes.f -- identical shape count, coordinates
// and colours -- drawn into an HTML <canvas> inside headless Chromium
// (see run_canvas_benchmark.py, which evaluates this file's `frame`
// function in the page). A real browser's Skia-backed 2D context, not a
// Node canvas shim.
//
// The canvas is NOT attached to the document and nothing is ever
// presented, matching the Festina side's offscreen surface: both are
// timed drawing into a buffer.
//
// The getImageData at the end is load-bearing rather than incidental.
// Skia batches and defers, so timing the loop alone can return before
// any pixel exists -- measured at 50.5 ms without this readback against
// 70.2 ms with it. It reads one pixel purely to force the flush, and it
// is inside the timed region on purpose.
function frame(ctx, shapes) {
    const t0 = performance.now();
    for (let i = 0; i < shapes; i++) {
        const x = (i * 37) % 780;
        const y = (i * 53) % 580;
        ctx.fillStyle = `rgb(${i % 255},${(i * 7) % 255},${(i * 13) % 255})`;
        ctx.fillRect(x, y, 12, 9);
        ctx.beginPath();
        ctx.arc(x + 6, y + 4, 4, 0, Math.PI * 2);
        ctx.fill();
    }
    ctx.getImageData(0, 0, 1, 1);
    return performance.now() - t0;
}
